### PR TITLE
Issue Fix

### LINKED ISSUE #112: [PE-D] add-appt : Exception is not thrown
I try the command add-appt pt/1 dr/2 at/2020-08-02 18:45 to/2020-08-02 19:00 dur/1H which should be an exception as it is specified in the user guide that exactly one of datetime_end and duration should be specified

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: godjuansan/ped#11

### LINKED ISSUE #116: [PE-D] edit-appt : edit duration not responsive
I try keying in edit-appt 1 dur/1H but nothing seems to change and there is no respond at all (not responsive)

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: godjuansan/ped#15

### LINKED ISSUE #125: [PE-D] Adding new appt right after end time of prev appt isnt allowed.
If an appt exists from 19:00 to 20:00, the next appointment cannot start at 20:00 and results in an error.

Perhaps could be an issue as appointments could start right after the previous appointment? 

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: jrvslam/ped#3

### LINKED ISSUE #132: [PE-D] Conflicts in time-slow of appointment not detected
When the doctor is booked by patient from start of month to end of month already, appointments within this month is still allowed for some reasons.

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: Hzxin/ped#1

### PR DESCRIPTION
**Pull request by Jacob-Pang**
** Changes **
* Changed EditAppointmentDescriptor to hold start, end and duration fields as the descriptor should only represent optional, non complex and independent fields. [ Fixes #116 ]
* Changed the criteria for Timeslot conflict [ Fixes #132 and #125 ]

** Testing **
* edit-appt taking in `at` field
image
image

* edit-appt taking in `to` field
image
image

* edit-appt taking in `dur` field
image
image

* add-appt with existing appt from 2020-2022 (all conflicting)
image

* add-appt with start time the same as existing appt end time
image
image

### CODE REVIEW

**Review by onnwards — APPROVED** [review]

**Comment by onnwards** [comment]
also fixes #112

File: src/main/java/seedu/address/logic/commands/appointment/EditAppointmentCommand.java [hunk 1]
```diff
@@ -116,8 +119,25 @@ public CommandResult execute(Model model) throws CommandException {
...
+            } else {
+                updatedTimeslot = new Timeslot(start, appointmentToEdit.getTimeslot().getEnd());
+            }
+        } catch (NegativeOrZeroDurationException e) {
+            throw new CommandException(Timeslot.MESSAGE_CONSTRAINTS);
+        }
+
+        // Tags
+        Set<Tag> updatedTags = editAppointmentDescriptor.getTags().orElse(appointmentToEdit.getTags());
+        Appointment editedAppointment = new Appointment(patientUuid, doctorUuid, updatedTimeslot, updatedTags);
```
Comment (icytornado) [inline src/main/java/seedu/address/logic/commands/appointment/EditAppointmentCommand.java:140]: LGTM

File: src/main/java/seedu/address/logic/commands/appointment/EditAppointmentCommand.java [hunk 2]
```diff
@@ -205,12 +216,28 @@ public void setDoctorIndex(Index doctorIndex) {
...
+        public void setEnd(LocalDateTime end) {
+            this.end = end;
+        }
+
+        public Optional<LocalDateTime> getEnd() {
+            return Optional.ofNullable(end);
+        }
+
+        public void setDuration(Duration duration) {
+            this.duration = duration;
```
Comment (icytornado) [inline src/main/java/seedu/address/logic/commands/appointment/EditAppointmentCommand.java:236]: LGTM

File: src/main/java/seedu/address/logic/parser/appointment/EditAppointmentCommandParser.java [hunk 3]
```diff
@@ -56,16 +56,19 @@ public EditAppointmentCommand parse(String args) throws ParseException {
...
+                    .setEnd(TimeslotParser.parseDateTime(argMultimap.getValue(PREFIX_TIMESLOT_END).get()));
         }
+
         if (argMultimap.getValue(PREFIX_TIMESLOT_DURATION).isPresent()) {
             editAppointmentDescriptor
-                    .setTimeslot(TimeslotParser
-                            .parseTimeslotByDuration(argMultimap.getValue(PREFIX_TIMESLOT_START).get(),
-                    argMultimap.getValue(PREFIX_TIMESLOT_DURATION).get()));
+                    .setDuration(TimeslotParser.parseDuration(argMultimap.getValue(PREFIX_TIMESLOT_DURATION).get()));
         }
```
Comment (icytornado) [inline src/main/java/seedu/address/logic/parser/appointment/EditAppointmentCommandParser.java:72]: LGTM

**Pull request merged**